tests/utils/retry/common.go

```diff
@@ -0,0 +1,28 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package retry
+
+import (
+	"context"
+	"fmt"
+	"time"
+)
+
+func waitAfterFail(ctx context.Context, retryWait time.Duration,
```

Comment:
I find all this decorating of contexts very difficult to follow.

Reply:
I understand you're not a fan of contexts (yet hehe!), I also like to keep it simple with signal channels.

But in this case we have different triggers that would cancel the context (aka close the channel), which can be the `cancel` function or the timer(s). Having these with a signal channel, multiple nested timers and select blocks would make it unreadable and needlessly complex. Even if context would not be part of std lib, I might consider using them in those cases 😉

Reply:
This function is called by `retry.UntilOk(ctx context.Context, poll time.Duration, func() (ok bool, err error)`.  If you made the signature of this function `retry.UntilOk(timeout time.Duration, poll time.Duration, func() (ok bool, err error)`, you could just handle the timeout and deadline internally without having to worry about passing in a context.   

In your tests you declare a global `ctx` before using this, and in the function you pass to `retry.UntilOK`, you add a timeout to that context using `context.WithTimeout`, and you do it again in `waitAfterFail`.  Not sure how this approach of decorating contexts from a parent to handle long running processes is easier to read.  I just want to reiterate, that contexts should be request scoped, and not for long running processes.  It would make sense to use a context to facilitate cancellation and deadline for an RPC call.  But cancellation shouldn't be confused with explicit stopping of long running services/goroutines.

Taken from Dave Cheney's blog: 
> However, context.Context‘s most important facility, broadcasting a cancellation signal, is incomplete as there is no way to wait for the signal to be acknowledged.
This is the problem when conflating cancellation and explicit stopping.  There's no way for us to know if services have actually stopped if you use contexts for stopping long running processes.

Reply:
> you could just handle the timeout and deadline internally without having to worry about passing in a context.

No because the top level context is used for nodes and is canceled when one of them crashes. So it's a mix of timeout and event driven, which would be very tedious to implement with channels and timers only, especially considering the different timeouts values here and there, there would be a lot more select cases all over the place.

> But cancellation shouldn't be confused with explicit stopping of long running services/goroutines.

At the end of the day, you do need a signal channel (internally or not) to signal to stop a long running goroutine. Contexts are just here to ease mixing event driven and time driven cancellation, without having to create a bunch of sub-signal-channels and sub-timers.

> is incomplete as there is no way to wait for the signal to be acknowledged.

Yes, but this is a bit out of scope; it is the same when closing a channel to stop a goroutine, you need another signal channel to wait for the goroutine to exit (and I agree this sucks too).
Such signal channels are implemented (or it's registered in `t.Cleanup`) in those tests.